Parity Signer

maciejhirsz opened the pull request:
Closes #1056

### Description

Introduces Parity Signer as another secure wallet option on the main screen. Account is unlocked by scanning a QR code of the address from the mobile app. Signing transactions is done in a modal overlaying the UI that first displays the QR code of the RLP of the transaction and then scans for the signature QR code from the mobile app.

### Changes

* New secure wallet has been added.
* New modal has been added to allow signing of the transaction while the `Promise` is being resolved.
* New reducers and actions have been added in order to pipe through the transaction from the UI to the `Promise` required by the `IFullWallet` interface.
* Added `@parity/qr-signer` as a new dependency.

### Steps to Test

1. Download the Parity Signer mobile app ([iOS](https://itunes.apple.com/us/app/parity-signer/id1218174838) / [Android](https://play.google.com/store/apps/details?id=com.nativesigner)). Create an account on it, send some ETH to the account (works with testnets).
1. Click on the "Parity Signer" option on the main screen.
1. Scan the account QR code from Parity Signer to unlock the wallet.
1. Create any kind of a transaction.
1. Sign the transaction using the app.
1. Make sure that the transaction goes through and is mined on chain.

Due to interactions with the webcam, integration tests might be tricky unless a lot of the stuff gets mocked away.

### TODO / Known issues

* [x] Needs a proper icon on the home screen.
* [x] `ParitySigner.tsx` needs to be moved to the containers.
* [x] UI needs some love (things are off-center, some paddings would be nice, some copy explaining what it is with links to the app would be nice).
* [x] Parity Signer needs a help website (linking to README.md for now...).

### Screenshots

![](http://maciej.codes/mycrypto/newicon.png)
![](http://maciej.codes/mycrypto/withlink.png)
![](http://maciej.codes/mycrypto/final3.png)

dternyak commented:
@maciejhirsz Are there any plans to support testnets on the parity mobile signer? 

Additionally, we'll want to disable Parity when non-ethereum networks are selected (e.g. disable parity option when network is ETC)

maciejhirsz commented:
Testnets are already supported, or did you run into issues? The UI doesn't separate accounts by network, but the chainId is correctly handled when signing the transaction.

Signing messages should be possible with the Parity Signer, I haven't investigated that yet, the UI (on the mobile app) might be a bit dumb about it, so will disable the panel there for now. Also will look at disabling it for non-Ethereum networks.

dternyak commented:
Got it. I didn't attempt a testnet tx as I didn't see a way to select testnet on the app, but it's good to see that's supported.

👍 to message signing as a later functionality.

maciejhirsz commented:
@dternyak I've been looking into it, and I don't really see any issues why the signer shouldn't be able to handle ETC and other networks as well. Basically anything that can be handled by keystore file / privatekey currently, should work with the app, unless there are some differences in how those networks sign txs/messages that I'm unaware of?

maciejhirsz commented:
Updated the PR:
* All UI stuff is fixed, latest `develop` merged back.
* Disabled Parity Signer for signing transactions (subject to future PR).
* Removed the **[WIP]** :tada:.

wbobeirne commented on common/components/Footer/index.tsx:
```diff
@@ -219,6 +220,7 @@ export default class Footer extends React.PureComponent<Props, State> {
           </div>
         </footer>
 
+        <QrSignerModal />
```
Would prefer this component only be rendered where it contextually makes sense, somewhere in the send form.

wbobeirne commented on common/components/WalletDecrypt/disables.ts:
```diff
@@ -22,9 +22,10 @@ export const DISABLE_WALLETS: { [key in WalletMode]: DisabledWallets } = {
+    wallets: [SecureWalletName.TREZOR, SecureWalletName.PARITY_SIGNER, MiscWalletName.VIEW_ONLY],
     reasons: {
       [SecureWalletName.TREZOR]: 'This wallet can’t sign messages',
+      [SecureWalletName.PARITY_SIGNER]: 'This wallet can’t sign messages',
```
Not your responsibility, but just a reminder that we should be translating this.

wbobeirne commented on common/containers/ParitySignerDecrypt/index.scss:
```diff
@@ -0,0 +1,14 @@
+        margin: 0 0.4em;
+    }
+
+    .qr-bounds {
```
We use [SuitCSS naming convention](https://github.com/MyCryptoHQ/MyCrypto/wiki/Style-Guides#classnames) to avoid class name conflicts. Likewise, the `a img` above would be better suited by class names.

wbobeirne commented on common/containers/ParitySignerDecrypt/index.tsx:
```diff
@@ -0,0 +1,67 @@
+        </div>
+        <p>{translate('ADD_PARITY_2')}</p>
+        <p>
+          <a href="https://itunes.apple.com/us/app/parity-signer/id1218174838" rel="nofollow">
```
Use the `NewTabLink` component here.

wbobeirne commented on common/containers/ParitySignerDecrypt/index.tsx:
```diff
@@ -0,0 +1,67 @@
+            styles={{ display: 'inline-block' }}
+          />
+        </div>
+        <p>{translate('ADD_PARITY_2')}</p>
```
At some point we should put together a help article that explains this process, or have some additional descriptive copy here.

maciejhirsz replied:
We are working on a proper help page that describes how to use the signer, I'll add a link to it here once it's done (ideally before publicity but can be done after the PR is merged).

wbobeirne commented on common/containers/ParitySignerDecrypt/index.tsx:
```diff
@@ -0,0 +1,67 @@
+    if (!isValidETHAddress(address)) {
+      console.error('Invalid address!');
+      return;
+    }
```
What's the user experience for when this happens? Does the signer keep watching for a potentially valid address?

maciejhirsz replied:
Yep, it will keep on scanning until a QR code with a valid address is presented.

wbobeirne commented on common/containers/QrSignerModal/index.scss:
```diff
@@ -0,0 +1,8 @@
+.QrSignerModal {
+  .qr-bounds {
```
Prefer SuitCSS naming convention.

wbobeirne commented on common/containers/QrSignerModal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+    // values in: wallets/non-deterministic/parity.ts
+    tx.v = Buffer.from([tx._chainId]);
+    tx.r = Buffer.from([0]);
+    tx.s = Buffer.from([0]);
```
Looks like you're mutating the `tx` props object here. If you want to alter something passed in as props, you'll need to copy it and save it in state. Something like...

```.tsx
public componentDidMount() {
  if (this.props.tx) {
    this.makeStateTx(this.props);
  }
}

public componentWillReceiveProps(nextProps: Props) {
  if (nextProps.tx && this.props.tx !== nextProps.tx) {
    this.makeStateTx(nextProps);
  }
}

private makeStateTx(props: Props) {
  const tx = { ...props.tx };
  tx.v = Buffer.from([tx._chainId]);
  tx.r = Buffer.from([0]);
  tx.s = Buffer.from([0]);
  this.setState({ tx });
}
```

maciejhirsz replied:
The values aren't read anywhere down the line, so it shouldn't really matter. Ideally there would be a way to serialize a transaction to RLP in ethereumjs without including the signature (`tx.serialize(false)` to match `tx.hash(false)`), but that's not the option. I can make a helper function for it here, and make a PR to ethereumjs.

wbobeirne replied:
The problem is it makes for non-deterministic redux state, makes it so there's no way to determine if the tx object has changed (without checking deep equality) and [in general is just advised against in React](https://reactjs.org/docs/components-and-props.html#props-are-read-only).

If you want to avoid the state boilerplate, it could at least do
```
const tx = { ...this.props.tx };
```
in `render`.

wbobeirne commented on common/libs/wallet/non-deterministic/parity.ts:
```diff
@@ -0,0 +1,59 @@
+        resolve(tx.serialize());
+      };
+
+      this.setWalletQrTransaction(tx, from, onSignature, onCancel);
```
Is there a reason this is a callback argument in constructor, instead of just resolving the promise with these variables? Would be nice if the wallet was more agnostic to where it gets used, so that you could use the same wallet instance to sign transactions with two different implementations.

maciejhirsz replied:
Yeah, this is the part I'm least happy about, and it goes along with how `QrSignerModal` is done. The callbacks go up to the container and are in turn placed in redux state for the modal to use.

The component that instantiates the wallet (`ParitySignerDecrypt` container in this case) has no knowledge of the modal, and is unmounted after unlocking so it cannot host the `QrSignerModal` itself, hence I'm passing these things over redux, which works but also feels over-engineered for my tastes.

Ideally I'd be able to pass the callbacks from the wallet to the modal as props without going through all these gymnastics, but I haven't been able to figure out how to make it do it any cleaner so far (my understanding of the existing code base is still pretty shallow).

wbobeirne replied:
We came up with a migration strategy to move these things to Redux via slack DMs. Those changes probably won't land for a little while though.

wbobeirne commented on common/reducers/wallet.ts:
```diff
@@ -30,9 +35,21 @@ export interface State {
 export const INITIAL_STATE: State = {
   inst: null,
   config: null,
+  signViaQr: null,
```
IMO (Would definitely like additional opinions) this should probably be its own reducer with its own set of actions / sagas. I think this is pretty separate from the other wallet processes, and is specific to only the parity signer.

dternyak replied:
I definitely see maintainability improvements with such a separation of concerns 👍

maciejhirsz commented on common/components/WalletDecrypt/WalletDecrypt.tsx:
```diff
@@ -177,7 +177,7 @@ const WalletDecrypt = withRouter<Props>(
         initialParams: {},
         unlock: this.props.setWallet,
-        helpLink: 'https://github.com/paritytech/parity-signer/blob/master/README.md'
+        helpLink: 'https://wiki.parity.io/Parity-Signer-Mobile-App-MyCrypto-tutorial.md'
```
[Yup!](https://github.com/paritytech/wiki/pull/54)
